### PR TITLE
Refactor extend in Containers

### PR DESCRIPTION
**Pull request by hyeok9855**
- [x] I've read the [.github/CONTRIBUTING.md](.github/CONTRIBUTING.md) file
- [x] My code follows the typing guidelines
- [x] I've added appropriate tests
- [x] I've run pre-commit hooks locally

## Description
Minor refactorings in subclasses of `Container`, which include:
1. Remove try-except error handling, as it seems unnecessary.
2. Handle the edge case when len(self) == 0 within the `extend` method instead of in the `__init__`.
3. Replace the padding function with an improved one for `Trajectories.extend`.

### CODE REVIEW

File: src/gfn/containers/states_container.py [hunk 1]
```diff
@@ -78,16 +78,7 @@ def __init__(
...
-        # self._log_rewards can be torch.Tensor of shape same as self.states or None.
-        if log_rewards is not None:
-            self._log_rewards = log_rewards
-        else:  # if log_rewards is None, there are two cases
-            if self.states.tensor.nelement() == 0:  # 1) initializing with empty states
-                self._log_rewards = torch.full(
-                    size=(0,), fill_value=-float("inf"), dtype=torch.float, device=device
-                )
-            else:  # 2) we don't have log_rewards and need to compute them on the fly
-                self._log_rewards = None
```
[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/containers/states_container.py:90]: Understood but we should ensure we don't lose any functionality - I don't clearly see how these three cases continue to be handled

File: src/gfn/containers/states_container.py [hunk 2]
```diff
@@ -145,14 +136,9 @@ def log_rewards(self) -> torch.Tensor:
...
             )
-            try:
-                self._log_rewards[self.is_terminating] = self.env.log_reward(
-                    self.terminating_states
-                )
-            except NotImplementedError:
-                self._log_rewards[self.is_terminating] = torch.log(
-                    self.env.reward(self.terminating_states)
-                )
+            self._log_rewards[self.is_terminating] = self.env.log_reward(
```
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/containers/states_container.py:139]: Cases are:
1. user overrides `Env.reward` -> should be okay since the default `Env.log_reward` simply returns `torch.log(self.reward(final_states))`
2. user overrides `Env.log_reward` -> also should be okay since it is implemented

The only case in which this change might cause an error is if the user overrides `Env.reward` and intentionally changes `Env.log_reward` as follows:
```python
    def log_reward(self, final_states: States) -> torch.Tensor:
        raise NotImplementedError
```
I don't think any reasonable user would do this, though.

[the remark this answers is not in the document]
Comment (josephdviviano) [reply src/gfn/containers/states_container.py:139]: I'm going to be annoying about this but I think we should support this case, there's no reason not to aside from making our base code cleaner.

File: src/gfn/containers/states_container.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/containers/states_container.py:90]: Those three cases were meant to avoid the edge case of **extending an empty container** (the second if).

I added this before when I encountered the edge case, but I recently realized it would be better to handle it directly in the `extend` method, which is exactly what I did in this PR.

**Comment by josephdviviano** [comment]
@hyeok9855 I'll leave it to your best judgement whether to take my idea or not - you've thought more deeply about this than me. I'm happy with either choice (we can fix it in the future if we need to).

File: src/gfn/containers/states_container.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hyeok9855) [reply src/gfn/containers/states_container.py:139]: > we should support this case

I don't really think so. I can't find any good reason why a user would go out of his way to change `Env.log_reward`.

> there's no reason not to aside from making our base code cleaner.

True, but making code cleaner is also important, I believe.

I'm pretty confident that this change won't cause any problems.

**Review by younik — APPROVED** [review]
I like this!
